### PR TITLE
attempt to solve routing problems first using transit of adjacent room

### PR DESCRIPTION
**Pull request by karlthepagan**
this adds the OCS room-by-room pathfinding to traveler as a fallback when things go wrong

### CODE REVIEW

**Comment by karlthepagan** [comment]
@chobobobo it was intentional: I use global.ROUTE_PRECALCULATION in order to avoid the undef error and make the parameter optional.

**Comment by karlthepagan** [comment]
I'm definitely hitting his case from reservers who arrive from the room to the west of this: E81S11

File: traveler.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ module.exports = function(globalOpts = {}){
         maxOps:            20000,
```
Comment (TooAngel) [inline traveler.js]: In the description is written `6 options` but there are actually 8.
And `roomRange` is missing in the description.

File: traveler.js [hunk 2]
```diff
@@ -130,12 +132,14 @@ module.exports = function(globalOpts = {}){
                 }
                 return matrix;
             };
-            return PathFinder.search(origPos, { pos: destPos, range: options.range }, {
+            const ret = PathFinder.search(origPos, { pos: destPos, range: options.range }, {
```
Comment (TooAngel) [inline traveler.js]: Define `ret` as const and change it afterwards `ret.route`. Is that fine for objects?

File: traveler.js [hunk 3]
```diff
@@ -231,8 +235,17 @@ module.exports = function(globalOpts = {}){
                     console.log(`TRAVELER: heavy cpu use: ${creep.name}, avg: ${travelData.cpu / travelData.count}, total: ${_.round(travelData.cpu, 2)}, pos: ${creep.pos}`);
                 }
                 if (ret.incomplete) {
-                    console.log(`TRAVELER: incomplete path for ${creep.name} from ${creep.pos} to ${destPos}`);
-                    if (ret.ops < 2000 && options.useFindRoute === undefined && travelData.stuck < gOpts.defaultStuckValue) {
+                    const route = ret.route && ret.route.length;
+                    console.log(`TRAVELER: incomplete path for ${creep.name} from ${creep.pos} to ${destPos}. Route length ${route}.`);
+                    if (route > 1) {
+                        ret = this.findTravelPath(creep, new RoomPosition(25, 25, ret.route[1].room),
```
Comment (TooAngel) [inline traveler.js]: I like that, we do that, too.

File: traveler.js (hunk 1 above)
Comment (karlthepagan) [inline traveler.js]: 👍

File: traveler.js (hunk 2 above)
Comment (karlthepagan) [inline traveler.js]: Is is permitted. Only frozen objects are immutable in ES6.

File: traveler.js (hunk 3 above)
Comment (karlthepagan) [inline traveler.js]: What target ranges have you tested? I'm preferring 22 (so 47 -> 3).

Comment (TooAngel) [inline traveler.js]: I chose 22, too. I know that someone played around with the value, not sure what was the conclusion.

Comment (karlthepagan) [inline traveler.js]: 24 leads to arriving on a swamp even with vision. I'd still entertain 23.

**Comment by chobobobo** [comment]
Unfortunately, for the room pathing issues I'm having no route is returned so this does not help.  Are there any cases you've found where an incomplete route is returned and this finds a route?

The way I'm reading it right now, this will simply get you to the next room on the route returned, but will ultimately end up in the same room that returned an incomplete route in the first place resulting in a dead end.

**Comment by karlthepagan** [comment]
With precalc on it seems reliable.

**Comment by karlthepagan** [comment]
Since this is going into traveler branch I'm going to merge. it's an improvement and hopefully gets traveler closer to dev.

**Pull request merged**